api/src/test/java/org/projectnessie/tools/catalog/migration/api/AbstractTestCatalogMigrator.java

```diff
@@ -0,0 +1,383 @@
+    Assertions.assertThat(result.registeredTableIdentifiers()).isEmpty();
+    Assertions.assertThat(result.failedToRegisterTableIdentifiers()).containsExactly(identifier);
+    Assertions.assertThat(result.failedToDeleteTableIdentifiers()).isEmpty();
+
+    // try to register same table twice which leads to AlreadyExistsException
+    result =
+        catalogMigratorWithDefaultArgs(deleteSourceTables)
+            .registerTables(Collections.singletonList(FOO_TBL2))
+            .result();
+    Assertions.assertThat(result.registeredTableIdentifiers()).containsExactly(FOO_TBL2);
+    Assertions.assertThat(result.failedToRegisterTableIdentifiers()).isEmpty();
+    Assertions.assertThat(result.failedToDeleteTableIdentifiers()).isEmpty();
+
+    result =
+        catalogMigratorWithDefaultArgs(deleteSourceTables)
+            .registerTables(Collections.singletonList(FOO_TBL2))
+            .result();
+    Assertions.assertThat(result.registeredTableIdentifiers()).isEmpty();
+    Assertions.assertThat(result.failedToRegisterTableIdentifiers()).contains(FOO_TBL2);
+    Assertions.assertThat(result.failedToDeleteTableIdentifiers()).isEmpty();
+  }
+
+  @ParameterizedTest
+  @ValueSource(booleans = {true, false})
+  public void testRegisterWithFewFailures(boolean deleteSourceTables) {
+    validateAssumptionForHadoopCatalogAsSource(deleteSourceTables);
+
+    // register only foo.tbl2
+    CatalogMigrationResult result =
+        catalogMigratorWithDefaultArgs(deleteSourceTables)
+            .registerTables(Collections.singletonList(FOO_TBL2))
+            .result();
+    Assertions.assertThat(result.registeredTableIdentifiers()).containsExactly(FOO_TBL2);
+    Assertions.assertThat(result.failedToRegisterTableIdentifiers()).isEmpty();
+    Assertions.assertThat(result.failedToDeleteTableIdentifiers()).isEmpty();
+
+    if (deleteSourceTables && !(sourceCatalog instanceof HadoopCatalog)) {
+      // create a table with the same name in source catalog which got deleted.
+      sourceCatalog.createTable(FOO_TBL2, schema);
+    }
+
+    // register all the tables from source catalog again. So that `foo.tbl2` will fail to register.
+    result = registerAllTables(deleteSourceTables);
+    Assertions.assertThat(result.registeredTableIdentifiers())
+        .containsExactlyInAnyOrder(FOO_TBL1, BAR_TBL3, BAR_TBL4);
+    Assertions.assertThat(result.failedToRegisterTableIdentifiers()).contains(FOO_TBL2);
+    Assertions.assertThat(result.failedToDeleteTableIdentifiers()).isEmpty();
+
+    Assertions.assertThat(targetCatalog.listTables(FOO))
+        .containsExactlyInAnyOrder(FOO_TBL1, FOO_TBL2);
+    Assertions.assertThat(targetCatalog.listTables(BAR))
+        .containsExactlyInAnyOrder(BAR_TBL3, BAR_TBL4);
+  }
+
+  @ParameterizedTest
+  @ValueSource(booleans = {true, false})
+  public void testRegisterNoTables(boolean deleteSourceTables) {
+    // use source catalog as targetCatalog which has no tables.
+    Assumptions.assumeFalse(
```

Comment:
It was because I made the logic complex for `testRegisterNoTables`. 

I simplified the logic and used `validateAssumptionForHadoopCatalogAsSource` itself.